url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/utils/Optimization.java

```diff
@@ -0,0 +1,332 @@
+package utils;
+
+import java.io.FileNotFoundException;
+import java.io.PrintStream;
+import java.util.Hashtable;
+import java.util.LinkedList;
+import java.util.List;
+
+import methodology.History;
+import protocols.Node;
+import protocols.State;
+import protocols.specifications.BestProtocol;
+import protocols.specifications.DBProtocol;
+import attributes.Attribute;
+import attributes.CryptoCalls;
+import attributes.DistanceFraudProbability;
+import attributes.FinalSlowPhase;
+import attributes.MafiaFraudProbability;
+import attributes.Memory;
+import attributes.SizeOfMessages;
+import attributes.TerroristFraudProbability;
+import attributes.relations.FinalSlowPhaseRelation;
+import attributes.relations.IntegerRelation;
+import attributes.relations.MemoryRelation;
+import attributes.relations.ProbabilityRelation;
+import attributes.relations.SizeOfMessagesRelation;
+import attributes.scales.KbitsScale;
+import attributes.scales.LogScale;
+import attributes.scales.NoScale;
+
+public class Optimization {
+
+	public static void main(String[] args) throws FileNotFoundException {
+		System.setOut(new PrintStream("optimization_reduced"+DBProtocol.MAX_N+".txt"));
+
+		createRoot();
+	}
+	
+	public static void createRoot(){
+		System.out.println("Starting");
+		//DBProtocol[][] protocols = DBProtocol.loadProtocolsFairly();
+		Attribute[] attributes = new Attribute[]{
+				new MafiaFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new DistanceFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new TerroristFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				//new TotalBitsExchanged(new IntegerRelation(), new NoScale<Integer>()),
+				//new TotalBitsExchanged(new BitsExchangedRelation(), new NoScale<Integer>()),
+				new SizeOfMessages(new SizeOfMessagesRelation(), new NoScale<Integer>()),
+				new CryptoCalls(new IntegerRelation(), new NoScale<Integer>()),
+				new Memory(new MemoryRelation(), new KbitsScale()),
+				new FinalSlowPhase(new FinalSlowPhaseRelation(), new NoScale<Boolean>()),
+		};
+		DBProtocol[] protocols = DBProtocol.loadProtocols();
+		System.out.println("Total protocols: "+protocols.length);
+		Node root = new Node(protocols);
+		System.out.println("Starting the optimization process");
+		List<DBProtocol>  nonDominated = getNonDpominated(root, protocols, attributes);
+		
+		System.out.println("Protocols dominating root = "+nonDominated.size());
+		
+		//search(root, protocols, attributes, new State[]{root.getState()});
+		//search(root, protocols, attributes);
+		DBProtocol.MAX_N = 2*DBProtocol.MAX_N;
+		protocols = DBProtocol.loadProtocols();
+		System.out.println("Total protocols: "+protocols.length);
+		searchByIteration(root, protocols, attributes);
+		
+	}
+
+	private static void search(Node root, DBProtocol[] protocols, Attribute[] attributes, State[] states){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 74-102 into a separate method.